File: 02desafios/desafio2.cpp
#include <bits/stdc++.h>
using namespace std;

bool resolve(int atual, string buf, int bpos, map<int, map<char,vector<int>>> tabtran, int* finais, char* alfabeto){
    if (bpos == buf.length() && finais[atual] == 1){
        return true;
    }else if(bpos == buf.length() || tabtran[atual][buf[bpos]].size()){
        return false;
    }else{
        for(int i=0;i<tabtran[atual][buf[bpos]].size();i++){
            if (resolve(tabtran[atual][buf[bpos]][i], buf, bpos+1, tabtran, finais, alfabeto)){
                return true;
            }
        }
        return false;
    }
}

int main(){
    int estados;
    scanf("%d",&estados);
    char alfabeto[1000];
    int simbolos;
    scanf("%d",&simbolos);
    for(int i=0;i<simbolos;i++)
        scanf(" %c",&alfabeto[i]);

    // int tabtran[estados]['z'+1];
    map<int, map<char,vector<int>>> tabtran;
    for(int i=0;i<estados;i++)
    {
        int pos,len; char trig;
        map<char,vector<int>> aux;
        for(int j=0;j<simbolos;j++)
        {
        scanf("%d %c %d",&pos,&trig,&len);
        vector<int> vetor;
        for(int j=0;j<len;++j){
            scanf("%d ", vetor[j]);
        }
        aux[trig] = vetor;
        tabtran[pos] = aux;
        // tabtran[i][trig]=vetor;
        }
    }
    int inicial;
    scanf("%d",&inicial);
    int finaisct;
    scanf("%d",&finaisct);
    int finais[estados];

    for(int i=0;i<estados;i++)
        finais[i]=-1;

    for(int i=0;i<finaisct;i++)
    {
        int final;
        scanf("%d",&final);
        finais[final]=1;
    }

    int atual=inicial;
    string buf;
    scanf(" %s",buf);
    int bpos=0;

    if(atual!=-1 && finais[atual]==1)
        printf("Aceito\n");
    else
        printf("Rejeito\n");
    return 0;
}

File: 02desafios/desafio2.py
def foo(atual, buf, bpos, tabtran, finais, alfabeto):
    if bpos == len(buf) and finais[atual] == 1:
        return True
    elif bpos == len(buf) or len(tabtran[atual][buf[bpos]]) == 0 or not(buf[bpos] in alfabeto):
        return False
    else:
        for i in range(len(tabtran[atual][buf[bpos]])):
            if foo(tabtran[atual][buf[bpos]][i], buf, bpos+1, tabtran, finais, alfabeto):
                return True
     

def main():
    #Leitura da quantidade de estados
    estados = int(input())

    #Leitura do alfabeto e da quantidade de simbolos
    alfabeto = [x for x in input().split(' ')]
    simbolos = int(alfabeto.pop(0))

    #Leitura da tabela de transição
    tabtran = {}
    # possibilidades = estados * simbolos
    i = 0
    for i in range(estados):
        tabtran[i] = {}
        for j in range(simbolos):
            #Leitura da posição atual, seu trigger
            #e o destino na tabela de transição
            st = input().split(' ')
            atual = st[0]
            trig = st[1]
            if not int(st[2]):
                prox = []
            else:
                prox = list(map(int,st[3:]))

            tabtran[i][trig] = prox

    #Leitura do Estado Inicial
    inicial = int(input())

    #Leitura dos Estados Finais
    finais = [-1 for x in range(estados)]
    st = input().split(' ')
    st.pop(0)
    i = 0
    while i < len(st):
        finais[int(st[i])] = 1
        i = i + 1

    #Leitura e verificação da string a ser verificada
    atual = inicial
    buf = input()
    bpos = 0

    if foo(atual, buf, bpos, tabtran, finais, alfabeto):
        print('Aceito')
    else:
        print('Rejeito')


main()
